### DISCUSSION TITLE
Manage API Keys used in concrete extensions

### DISCUSSION THREAD

**Discussion by Carlosmape** [Ulauncher/Ulauncher#1030] [Extensions and Themes]
Hi all!

A few weeks ago I started to write a Ulauncher extension and published it in to the extensions repository. The problem is that I would like to used a Google API key to integrame some google service (in this case Google Calendar integration), and obviously pushing a file containing the API Key into a public github repository is not a good practice.

Taking into account that extensions are downloaded directly from a public Github repository. It is needed to include the key with the source code. It is a must (it has to sense to publish an extension that integrates with any API without it API key or token)

In this way, there are some ideas that comes to me: 

First of all, make a kind of (bidirectional) hash to make the key (or key file) not readable. My extension should de-hash the key and then use it, however the hashing method would be published inside the extension source (and anyone could use it to decode mine). Not a good idea at all.

Secondly, I tough that Ulauncher could include some commonly used API keys (the Google one among others) and give some method to developers to use Ulauncher's keys. But this feature will be (I guess) just for Ulauncher installed from binary only, because it could not be published in the Ulauncher repository (as in my concrete case) and Ulauncher installed from source code shall not count with this token file. So I don't know if it is feasible...

Did anyone think about this topic? I searched in the web for another solutions but I can't find a properly way to solve the problem without ignore the key file from my repository.

Regards!

**Comment by friday** [comment]
You would have to make it a user-setting (stored in plain text) or put up some kind of proxy for their API that your extension users can use, but the latter would cost you money.

Better to try to integrate with [GOA](https://wiki.gnome.org/Projects/GnomeOnlineAccounts) ([API documentation](https://developer-old.gnome.org/goa/stable/))

I believe GOA works on all desktops. But people not using Gnome might be reluctant to use/install it.

**Reply by Carlosmape** [reply]
I think that to make the end-user create it own key and then paste it into the extension settings is a little weird... In fact, I used [Google OAuth](https://developers.google.com/identity/protocols/oauth2) method to avoid this problem, and let the user simply grant permissions to my extension.

GOA seems to be a good approach however I would prefer the extension to be stand-alone (without any extra software dependencies).

**Reply by Carlosmape** [reply]
Update:
I used API key in preferences approach and got:
```
google.auth._default: _get_gce_credentials() | Authentication failed using Compute Engine authentication due to unavailable metadata server.
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/ulauncher/api/client/Client.py", line 54, in on_message
    self.extension.trigger_event(event)
  File "/usr/lib/python3.10/site-packages/ulauncher/api/client/Extension.py", line 52, in trigger_event
    action = listener.on_event(event, self)
  File "/home/slimbook/.local/share/ulauncher/extensions/com.github.carlosmape.ulauncher-calendar/extension/KeywordQueryEventListener.py", line 17, in on_event
    extension_result = extension.GetExtensionResult(query, int(max_ev))
  File "/home/slimbook/.local/share/ulauncher/extensions/com.github.carlosmape.ulauncher-calendar/extension/AlmanacExtension.py", line 40, in GetExtensionResult
    self.calendar.initialize_google()
  File "/home/slimbook/.local/share/ulauncher/extensions/com.github.carlosmape.ulauncher-calendar/extension/calendar_manager.py", line 29, in initialize_google
    self.google = build('calendar', 'v3', developerKey = self.api_key)
  File "/home/slimbook/.local/lib/python3.10/site-packages/googleapiclient/_helpers.py", line 131, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/home/slimbook/.local/lib/python3.10/site-packages/googleapiclient/discovery.py", line 298, in build
    service = build_from_document(
  File "/home/slimbook/.local/lib/python3.10/site-packages/googleapiclient/_helpers.py", line 131, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/home/slimbook/.local/lib/python3.10/site-packages/googleapiclient/discovery.py", line 576, in build_from_document
    credentials = _auth.default_credentials(
  File "/home/slimbook/.local/lib/python3.10/site-packages/googleapiclient/_auth.py", line 54, in default_credentials
    credentials, _ = google.auth.default(scopes=scopes, quota_project_id=quota_project_id)
  File "/usr/lib/python3.10/site-packages/google/auth/_default.py", line 488, in default
    raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started
```   
I tried to call the api using raw post query and it returned:
```
{
  "error": {
    "code": 401,
    "message": "API keys are not supported by this API. Expected OAuth2 access token or other authentication credentials that assert a principal. See https://cloud.google.com/docs/authentication",
    "errors": [
      {
        "message": "Login Required.",
        "domain": "global",
        "reason": "required",
        "location": "Authorization",
        "locationType": "header"
      }
    ],
    "status": "UNAUTHENTICATED",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "reason": "CREDENTIALS_MISSING",
        "domain": "googleapis.com",
        "metadata": {
          "method": "calendar.v3.Events.Insert",
          "service": "calendar-json.googleapis.com"
        }
      }
    ]
  }
}
```

And effectively if you refer to Google Calendar API: 
 
![image](https://user-images.githubusercontent.com/5487555/167915788-83ab146c-ca36-47a7-b076-dbcd66f6f760.png)

**Comment by friday** [comment]
Some problems just can't be solved in the way you want. The same situation would apply to us if we wanted to support installing extensions from private github repos.

Only reasonable solution is GOA imo, but either way this is not a Ulauncher concern or Ulauncher extension development concern. You'd be better off asking on a programming forum/chat/site like stack overflow.

**Reply by Carlosmape** [reply]
That's it. This is because I write here. To expose a (I think) trouble that could affect some other extension developer. So:

- Is Ulauncher compatible with private repositories?
- Will Ulauncher API handle it owns 3th party API integrations? (Because of your answers I assume that no)

As you say, in my concrete case, to solve the problem I should change the approach (integrate with other things like Trello or something like this) or ask in Google forums. Thanks for your responses, @friday
